### PR TITLE
[RNMobile] Reimplementation of block borders

### PR DESCRIPTION
**Pull request by dratwas**
## Description
In this PR i reimplemented borders of blocks according to that issue https://github.com/wordpress-mobile/gutenberg-mobile/issues/1948

Gutenberg-mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2050

I removed old borders and calculations related to them and set borders as a view that is positioned absolutely.

I also had to manage all margins because they changed after removing borders. I mentioned a few issues related to that [here](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1948#issuecomment-597574611). I fixed them by using negative margins in the parent (FlatList) except the root list.
So the base concept is - if the parent has margin set to -16 and each inner block has margin set to 16 then the edge margins are cut and we don't need to check if it is first/last block etc. This solution should work for horizontal blocks as well w/o any calculations. Tested on columns and buttons.

I also added a possibility to change the horizontal or vertical margin of inner blocks since @lukewalczak said he would need that probably in buttons block.

After these changes, I had to fix margins in cover, media, and group block. 

NOTE:
This PR needs an accurate review of spacing of borders and inner blocks. @iamthomasbishop could you please check it? :)

## How has this been tested?
All tests cases related to the spacing and borders from here:
https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md

Basically the editor should look the same as before :)

## Types of changes
Reimplementation of blocks borders

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-editor/src/components/block-list/index.native.js [hunk 1]
```diff
@@ -99,14 +101,22 @@ export class BlockList extends Component {
...
+			marginHorizontal = styles.defaultBlock.marginLeft,
 		} = this.props;
 
 		const forceRefresh =
 			shouldShowInsertionPointBefore || shouldShowInsertionPointAfter;
 
+		const containerStyle = {
+			flex: isRootList ? 1 : 0,
+			marginVertical: isRootList ? 0 : -marginVertical,
+			marginHorizontal: isRootList ? 0 : -marginHorizontal,
```
Comment (pinarol) [inline packages/block-editor/src/components/block-list/index.native.js]: It can be good to drop some code comments here about why this was needed.

**Review by pinarol — COMMENTED** [review]
Masterpiece! Seriously! 🎉 Can't describe how happy I am to see lots of code removed.

I have detected 3 issues among below tests but in general I think no big issues here. I only tested with iOS. 
@jbinda Could you test with Android as well?

- [x] Media Text block - Media & Text alignment
- [x] Media Text block - Vertical alignment

- [x] Group - AppenderButton is rendered - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc001)
- [x] Group - Deep nesting is possible (iOS only) - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc002)
- [x] Group - Check if Group placeholder is visible for the unselected state - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc003)
- [x] Group - Check if Group placeholder is render in nested structure - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc004)
- [x] Group - Nested block have proper border styling - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc005)
- [x] Group - Nested block have proper margins values - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc006)
- [x] Group - Breadcrumbs on FloatingToolbar is properly displayed - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc008)
- [x] Group - Navigation up button works as expected - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc009)
- [x] Group - Navigation down works according to parent-first approach - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc010)
- [x] Group - Cross navigation between blocks works as expected - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc011)
- [x] Group - Ungroup button works as expected - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc012)
- [x] Group - Check if in DarkMode all components gets proper colors (iOS) - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc013)
- [x] Group - Check if nested Placeholder block can be replaced - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc014)  ❌ <<< I saw the issue in [this PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/1984) while testing it, was that fix merged into this branch?

- [x]  The empty space under toolbar, see [code comments](https://github.com/WordPress/gutenberg/pull/20769#discussion_r397342311)

- [x]  Media-Text: The extra space between media - text, see [code comments](https://github.com/WordPress/gutenberg/pull/20769#discussion_r397348503)

File: packages/block-editor/src/components/block-list/index.native.js [hunk 2]
```diff
@@ -118,7 +128,10 @@ export class BlockList extends Component {
 					innerRef={ this.scrollViewInnerRef }
 					extraScrollHeight={ innerToolbarHeight + 10 }
 					keyboardShouldPersistTaps="always"
-					scrollViewStyle={ { flex: isRootList ? 1 : 0 } }
+					scrollViewStyle={ {
+						flex: isRootList ? 1 : 0,
+						overflow: 'visible',
```
[the remark this answers is not in the document]
Comment (jbinda) [reply packages/block-editor/src/components/block-list/index.native.js]: as far as I remember from the conversation with Piotr the `overflow: 'visible'` is to prevent edge to be cut is some cases. In example without it in Column PR borders was cropped.

However I didn't tested the latest version and if you do not noticed any difference by removing this I will try to confirm that on Android as well and let you know.

If not the possible solution is to remove `overflow: 'visible'`  only for root list - we already have a prop called `isRootList`

**Comment by jbinda** [comment]
@pinarol sure I will test in on Android today in 1st place

**Comment by pinarol** [comment]
I can't repro this anymore, it should have been fixed by the latest merge from master by [this PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/1984).

- [x] Group - Check if nested Placeholder block can be replaced

File: packages/block-editor/src/components/block-list/index.native.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbinda) [reply packages/block-editor/src/components/block-list/index.native.js]: Targeting current master branch it is ok to remove that prop. I double check and borders for currently available blocks seems to work as expected without it. 

Will test in Columns PR as well later. 

However we discussed with Piotr that we can remove it from this PR and if necessary add in [Column PR adjustment](https://github.com/WordPress/gutenberg/pull/21073). That will make more sense because according to above it is adjustments for Columns

File: packages/block-editor/src/components/block-list/block.native.js [hunk 3]
```diff
@@ -79,104 +79,6 @@ class BlockListBlock extends Component {
 		);
 	}
 
-	applySelectedBlockStyle() {
```
[the remark this answers is not in the document]
Comment (dratwas) [reply packages/block-editor/src/components/block-list/block.native.js]: `isAncestorSelected` is used to determine if the block should be dimmed but I removed the rest, thanks :)

**Comment by jbinda** [comment]
Tested on Android demo and WP Android:

- [x] Media Text block - Media & Text alignment
❓ Media Text block - Vertical alignment - please see below question
- [x] Group - AppenderButton is rendered - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc001)
- [x] Group - Deep nesting is possible (iOS only) - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc002)
- [x] Group - Check if Group placeholder is visible for the unselected state - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc003)
- [x] Group - Check if Group placeholder is render in nested structure - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc004)
- [x] Group - Nested block have proper border styling - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc005)
- [x] Group - Nested block have proper margins values - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc006)
- [x] Group - Breadcrumbs on FloatingToolbar is properly displayed - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc008)
- [x] Group - Navigation up button works as expected - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc009)
- [x] Group - Navigation down works according to parent-first approach - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc010)
- [x] Group - Cross navigation between blocks works as expected - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc011)
- [x] Group - Ungroup button works as expected - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc012)
- [x] Group - Check if in DarkMode all components gets proper colors (iOS) - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc013)
- [x] Group - Check if nested Placeholder block can be replaced - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/group.md#tc014) 



**Media Text block - Vertical alignment - question**

Is below screen present proper item alignment for Horizontal layout when vertical alignment is set to top ?

<image src="https://user-images.githubusercontent.com/21242757/77534676-61516f80-6e99-11ea-8df9-477898ebf94a.png" width="500" />

Apart of above checklist I have noticed another thing with Separator line. Depends on hierarchy level the Separator line has different width. Not sure if it is expected or something brings by this PR. However maybe it is worth to fix this here to make Separator behaviour consistent with other changes. This is question to @iamthomasbishop probably but shouldn't the Separator line align to content as well  ?

 See below screens for details.

File: packages/block-editor/src/components/block-list/index.native.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (dratwas) [reply packages/block-editor/src/components/block-list/index.native.js]: I removed it since it is not needed at this stage. I think if it will be needed for columns, we should add it in columns PR.

**Comment by dratwas** [comment]
@jbinda 
>Is below screen present proper item alignment for Horizontal layout when vertical alignment is set to top ?

I think so, apart from the space on left and right (already fixed)

**Comment by pinarol** [comment]
> Is below screen present proper item alignment for Horizontal layout when vertical alignment is set to top ? (the dashed border frame is not align with image)

> Depends on hierarchy level the Separator line has different width. 

Those 2 work the same way on prod. If there's something that needs to be improved about these items we can handle it on a different PR since this PR's purpose is only improving the code quality while making sure that we aren't breaking anything.

File: packages/block-library/src/media-text/style.native.scss [hunk 4]
```diff
@@ -104,27 +108,35 @@
...
-	padding: $block-edge-to-content;
+// Media
+.mediaLeft {
+	padding-right: $block-edge-to-content;
 }
 
-.denseMediaPadding {
-	padding: $block-media-container-to-content;
+.mediaRight {
+	padding-left: $block-edge-to-content;
```
[the remark this answers is not in the document]
Comment (dratwas) [reply packages/block-library/src/media-text/style.native.scss]: Fixed :)

File: packages/block-editor/src/components/block-list/block.native.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (pinarol) [reply packages/block-editor/src/components/block-list/block.native.js]: cool, I missed that, thanks!

**Comment by jbinda** [comment]
Cool ! That was the only findings that I was able to spot during my todays testing

**Review by pinarol — APPROVED** [review]
I repeated my tests on iOS, issues I reported are fixed. I can't detect any regressions this PR is causing. I think this is ready to go. Thank you for all the great work! Again, masterpiece!

@jbinda please give 👍 for Android side if you  are finished with testing.

File: packages/block-editor/src/components/block-list/block.native.scss [hunk 5]
```diff
@@ -134,6 +59,28 @@
 }
 
 .neutralToolbar {
-	margin-left: -$block-edge-to-content;
-	margin-right: -$block-edge-to-content;
+	margin-left: - $block-edge-to-content;
+	margin-right: - $block-edge-to-content;
```
Comment (jbinda) [inline packages/block-editor/src/components/block-list/block.native.scss]: not relevant however it seems like an extra space here

**Review by jbinda — APPROVED** [review]
I have tested on Android and also do not noticed any regression.

Let's ship it :)

**Comment by jbinda** [comment]
hooray 🚀
